rclcpp/include/rclcpp/service.hpp

```diff
@@ -121,6 +121,17 @@ class ServiceBase
   bool
   exchange_in_use_by_wait_set_state(bool in_use_state);
 
+  /// Add the Service to a wait set.
+  /**
+   * \param[in] wait_set A handle to the wait set to add the Service to.
+   * \return `true` if the Service is added successfully, `false` otherwise.
+   * \throws rclcpp::execptions::RCLError from rcl_wait_set_add_*()
+   */
+  RCLCPP_PUBLIC
+  virtual
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) = 0;
```

Comment:
My one concern with this new API is how it will interact with the existing [`WaitSet` classes](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/wait_set.hpp). I suppose this will not work as expected if someone chooses to use a `WaitSet` object. Maybe we should update this statement to use the new API:

https://github.com/ros2/rclcpp/blob/2d6e6364cda34af4c8d005f5c7dde0bdab4a8d78/rclcpp/include/rclcpp/wait_set_policies/detail/storage_policy_common.hpp#L362-L365

I'm not sure though; I'd like to hear from @ivanpauno and @wjwwood

Reply:
Yeah, this new method seems like a poor way to control whether or not a service is handled in the managed node, and it won't work with the WaitSet classes.

Reply:
I guess we don't have lifecycle subscriptions yet either, but I would say we need a new way to indicate whether a subscription should be waited on or not.

Initially I thought the callback groups could already be used for this, but you could have a callback group with one normal service and one lifecycle service. So I don't think that will be useful.

The other question is, should it just not "get execution time" as was quoted in the first comment of this pr, or should requests received while not active be discarded? I.e., imagine: "service is inactive" -> "request is received" -> "service becomes active", should the request received before being active be handled? My intuition is no, but I'm not 100% sure.

If we said that requests received during an inactive state are ignored, we should implement the lifecycle service using a callback that wraps the user's callback and only calls it if the lifecycle state is active.

Reply:
@wjwwood  Hmm... ignoring requests in inactive state seems fine too. If we decide to ignore requests, what do you think about filtering by overriding the handle request instead of wrapping the callback? (The initial implementation did that.)

@alsora what do you think about ignoring requests in inactive state?

Reply:
According to the design document https://design.ros2.org/articles/node_lifecycle.html

> While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
In the inactive state, any data that arrives on managed topics will not be read and or processed. Data retention will be subject to the configured QoS policy for the topic.

This seems to only refer to execution time, i.e. the executor will ignore the entity for as long as it is inactive.

From a performance point of view, the current approach is also much more efficient than having a "wrapper callback".
The  "wrapper callback" approach requires deserialization, going through all the ROS layers and doing work in the executor.
When a  server is inactive, it may be for a  variety of reason, but I wouldn't want  it to perform any work at all.

Lastly, although a wrapper callback may work with subscriptions, how would that be implemented with servers where the callback usually needs to send a response back to the client?
Will this require for the response to have a "success/failure"  bool field ? How does the client interpret if the server invocation failed due to the server being inactive or due to problems in the request?

For what concerns the fact that this won't work with the new WaitSet class, is this really a problem? The point of the WaitSet class is to  decouple waiting and scheduling. The user is responsible for taking the data and executing it, so it should be easy to handle this.
Anyhow, can you elaborate why this wouldn't work? The WaitSet class is currently calling `rcl_wait_set_add_service`, can't we have it call `server->add_to_wait_set()` API which is overridden here?

For what is worth, I looked into how to implement lifecycle entities also with the RMW listener APIs and the events executor. Here the approach would be slightly different since the `add_to_wait_set` function is not invoked.
The approach would consist in the `on_activate` and `on_deactivate` functions to remove/restore the listener callback if present.

Reply:
> If we decide to ignore requests, what do you think about filtering by overriding the handle request instead of wrapping the callback? (The initial implementation did that.)

That could work, but maybe it produces overhead you don't want, as @alsora said.

>  When a server is inactive, it may be for a variety of reason, but I wouldn't want it to perform any work at all.

I think to achieve this we would need to have a new active state in each entity and the executor/wait set in rclcpp would need to know to ignore it when it is inactive, as well we'd need a way to notify the executor/wait set when that state changes (so it can build a new wait set that includes it).

Once we have that, we can either let the messages/requests accumulate or we can clear them as we become active, which ever seems appropriate (we could even configure them for that).

> Lastly, although a wrapper callback may work with subscriptions, how would that be implemented with servers where the callback usually needs to send a response back to the client?
Will this require for the response to have a "success/failure" bool field ? How does the client interpret if the server invocation failed due to the server being inactive or due to problems in the request?

I would say the request is taken (cleared out), but never acted on. This would be no different than any other timed out service call from the client's perspective. This is already the case and a shortcoming of services, so it's something worth improving, but not specific to this issue (lifecycle behavior), as there are other reasons a service may appear available to the client but fail to respond (service callback throws or service server is destroyed before request is received, etc).

Reply:
> For what concerns the fact that this won't work with the new WaitSet class, is this really a problem? The point of the WaitSet class is to decouple waiting and scheduling. The user is responsible for taking the data and executing it, so it should be easy to handle this.
Anyhow, can you elaborate why this wouldn't work? The WaitSet class is currently calling rcl_wait_set_add_service, can't we have it call server->add_to_wait_set() API which is overridden here?

It can work, the `WaitSet` class just needs to be updated. However, I would not do it by overriding the `add_to_wait_set()` method. I would instead add a state to all waitable entities like "active" and let the wait set introspect that when deciding whether or not to add it to the rcl wait set.

Plus as I said before, you need a feedback mechanism so that when that state changes the wait set can wake up and rebuild the wait set to include the newly active items.